From 247a02fe1816d280f7b57a07e590d9308fa46bfc Mon Sep 17 00:00:00 2001
From: Ana Krivokapic <akrivoka@redhat.com>
Date: Thu, 11 Jul 2013 12:50:01 +0200
Subject: [PATCH] Make sure replication works after DM password is changed

Replica information file contains the file `cacert.p12` which is protected by
the Directory Manager password of the initial IPA server installation. The DM
password of the initial installation is also used for the PKI admin user
password.

If the DM password is changed after the IPA server installation, the replication
fails.

To prevent this failure, add the following steps to ipa-replica-prepare:
1. Regenerate the `cacert.p12` file and protect it with the current DM password
2. Update the password of the PKI admin user with the current DM password

https://fedorahosted.org/freeipa/ticket/3594
---
 freeipa.spec.in                   |  7 +++++--
 install/tools/ipa-replica-prepare | 39 +++++++++++++++++++++++++++++++++++++++
 2 files changed, 44 insertions(+), 2 deletions(-)

diff --git a/freeipa.spec.in b/freeipa.spec.in
index e253da1ad4b1c41db9cdccd97d48d90f9f0fe658..70ed1dfd431d20fedfdc19aa1b00fcfe0920876a 100644
--- a/freeipa.spec.in
+++ b/freeipa.spec.in
@@ -24,7 +24,7 @@ Source0:        freeipa-%{version}.tar.gz
 BuildRoot:      %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
 
 %if ! %{ONLY_CLIENT}
-BuildRequires: 389-ds-base-devel >= 1.2.11.14
+BuildRequires:  389-ds-base-devel >= 1.3.1.3
 BuildRequires:  svrcore-devel
 BuildRequires:  /usr/share/selinux/devel/Makefile
 BuildRequires:  policycoreutils >= %{POLICYCOREUTILSVER}
@@ -106,7 +106,7 @@ Requires: %{name}-python = %{version}-%{release}
 Requires: %{name}-client = %{version}-%{release}
 Requires: %{name}-admintools = %{version}-%{release}
 Requires: %{name}-server-selinux = %{version}-%{release}
-Requires(pre): 389-ds-base >= 1.2.11.14-1
+Requires(pre): 389-ds-base >= 1.3.1.3
 Requires: openldap-clients
 Requires: nss
 Requires: nss-tools
@@ -814,6 +814,9 @@ fi
 %ghost %attr(0644,root,apache) %config(noreplace) %{_sysconfdir}/ipa/ca.crt
 
 %changelog
+* Wed Jul 10 2013 Ana Krivokapic <akrivoka@redhat.com> - 3.2.99-4
+- Bump minimum version of 389-ds-base to 1.3.1.3 for user password change fix.
+
 * Wed Oct 17 2012 Martin Kosek <mkosek@redhat.com> - 2.99.0-51
 - Print ipa-upgradeconfig errors during RPM update
 
diff --git a/install/tools/ipa-replica-prepare b/install/tools/ipa-replica-prepare
index 6d01c4c0017ab7983d3cea6dd6b1142dd8002fea..16dceaa16a80b0de8e4d8e02082f61c7c3c136ed 100755
--- a/install/tools/ipa-replica-prepare
+++ b/install/tools/ipa-replica-prepare
@@ -29,6 +29,7 @@ import krbV
 from ipapython import ipautil
 from ipaserver.install import bindinstance, dsinstance, installutils, certs
 from ipaserver.install.bindinstance import add_zone, add_fwd_rr, add_ptr_rr, dns_container_exists
+from ipaserver.install.cainstance import DEFAULT_DSPORT
 from ipaserver.install.replication import enable_replication_version_checking
 from ipaserver.install.installutils import resolve_host, BadHostError, HostLookupError
 from ipaserver.plugins.ldap2 import ldap2
@@ -238,6 +239,39 @@ def copy_files(realm_name, dir):
 def get_dirman_password():
     return installutils.read_password("Directory Manager (existing master)", confirm=False, validate=False)
 
+def update_pki_admin_password(dirman_password):
+    ldap = ldap2(shared_instance=False,
+                 ldap_uri='ldaps://%s:7390' % api.env.host,
+                 base_dn=DN('o=ipaca'))
+    ldap.connect(
+        bind_dn=DN(('cn', 'directory manager')),
+        bind_pw=dirman_password
+    )
+    dn = DN('uid=admin', 'ou=people', 'o=ipaca')
+    ldap.modify_password(dn, dirman_password)
+    ldap.disconnect()
+
+def regenerate_ca_file(dirman_password, ca_file):
+    dm_pwd_fd = ipautil.write_tmp_file(dirman_password)
+
+    keydb_pwd = ''
+    with open('/var/lib/pki-ca/conf/password.conf') as f:
+        for line in f.readlines():
+            key, value = line.strip().split('=')
+            if key == 'internal':
+                keydb_pwd = value
+                break
+
+    keydb_pwd_fd = ipautil.write_tmp_file(keydb_pwd)
+
+    ipautil.run([
+        '/usr/bin/PKCS12Export',
+        '-d', '/var/lib/pki-ca/alias/',
+        '-p', keydb_pwd_fd.name,
+        '-w', dm_pwd_fd.name,
+        '-o', ca_file
+    ])
+
 def main():
     installutils.check_server_configuration()
     options, args = parse_options()
@@ -351,6 +385,11 @@ def main():
             if not certs.ipa_self_signed():
                 # FIXME, need option for location of CA backup
                 if ipautil.file_exists(options.ca_file):
+                    # Since it is possible that the Directory Manager password
+                    # has changed since ipa-server-install, we need to regenerate
+                    # the CA PKCS#12 file and update the pki admin user password
+                    regenerate_ca_file(dirman_password, options.ca_file)
+                    update_pki_admin_password(dirman_password)
                     shutil.copy(options.ca_file, dir + "/cacert.p12")
                 else:
                     raise RuntimeError("Root CA PKCS#12 not found in %s" % options.ca_file)
-- 
2.7.4

